### PR TITLE
add a deadlock test

### PR DESCRIPTION
**Pull request by juchiast**
@realcr @kamyuentse 
Hey, I have a deadlock here. The key point is polling `task2` before `task1`.

### CODE REVIEW

**Comment by juchiast** [comment]
I try to fix the deadlock by introducing `NotPolled` state in `AcquireFutureState`. With this change, `acquire()` will do nothing, it'll just return an `AcquireFuture` in `NotPolled` state. Code in old `acquire()` is moved into `poll()`.

This change messed up `simple` and `multiple` test, which I think is not correctly written. `Core::run()` only waits for the passed future to complete, it does not wait for all spawned futures to complete.

I think we might need to rewrite tests. I'm not familiar with writing `futures` and `tokio`. What do you think about this?

**Comment by realcr** [comment]
@juchiast : I'm checking it.

**Comment by realcr** [comment]
@juchiast: Hi, what's up? 
Don't worry about being new to `futures` and `tokio`. You were assigned this issue so that you can experiment with it. I also think that it was a good idea to introduce a test that fails. This way we can help you out solve the problem.

I first checked out commit: https://github.com/realcr/async_mutex/pull/9/commits/5518db0226b1079e09baf13b12ebc34f0a022ab0 , just before you added the NotPolled state.

Next, I added some prints to see what happens:
```rust
    #[test]
    fn deadlock() {
        use tokio::timer::Deadline;
        use std::time::Duration;
        use std::time::Instant;

        let mut core = Core::new().unwrap();

        let async_mutex = AsyncMutex::new(NumCell { num: 0});

        let task0 = async_mutex.acquire(|mut num_cell| -> Result<_, (_, ())> {
            println!("Enter task0");
            num_cell.num += 1;
            println!("Leave task0");
            Ok((num_cell, ()))
        });

        let task1 = async_mutex.acquire(|mut num_cell| -> Result<_, (_, ())> {
            println!("Enter task1");
            num_cell.num += 1;
            println!("Leave task1");
            Ok((num_cell, ()))
        });

        let task2 = async_mutex.acquire(|mut num_cell| -> Result<_, (_, ())> {
            println!("Enter task2");
            num_cell.num += 1;
            println!("Leave task2");
            Ok((num_cell, ()))
        });

        fn timeout() -> Instant {
            Instant::now() + Duration::from_secs(1)
        }

        core.run(task0).unwrap();
        println!("Done task0!");
        core.run(Deadline::new(task2, timeout())).unwrap();
        core.run(Deadline::new(task1, timeout())).unwrap();
    }
```

This is what I got:

```bash
$ cargo test deadlock
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running target/debug/deps/async_mutex-4fef7cbd0580d665

running 1 test
test tests::deadlock ... FAILED

failures:

---- tests::deadlock stdout ----
Enter task0
Leave task0
Done task0!
thread 'tests::deadlock' panicked at 'called `Result::unwrap()` on an `Err` value: DeadlineError(Elapsed)', libcore/result.rs:945:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

So, initially we tried to acquire in this order: `task0, task1, task2`
According to the prints, we did manage to enter task0 and leave it successfully, so task0 is out of the picture. The waiters for the resource are now `task1, task2`.

Next, we have this code:

```rust
core.run(Deadline::new(task2, timeout())).unwrap();
core.run(Deadline::new(task1, timeout())).unwrap();
```

The first line says: I will wait for task2 to finish, or cancel with a deadline error. task2 can not obtain the resource because it is waiting for task1, which is the correct behaviour. Therefore after a while the first line returns a deadline error which is unwrapped.

To fix the problem, I think that we should swap the two lines, to be:
```rust
core.run(Deadline::new(task1, timeout())).unwrap();
core.run(Deadline::new(task2, timeout())).unwrap();
```
If we do that, the test passes. However, I'm not sure we are testing something very interesting this way. What did you originally intend to test with the deadlock test?

Some other thing: Note that almost none of our test have a timer inside of it. This is on purpose. Time is not very deterministic, and using time inside a test could give non deterministic results (The test sometimes passes and sometimes fails). For the same reason we try to avoid having real random inside tests.
The other side of this is that if we make the time intervals longer we can get more accurate results, at the cost of having a slow test.

If you want to design a special test that checks some special order of events, the proper way to do it is probably to use messages (For example, with a bunch of oneshots: https://docs.rs/futures/0.1/futures/sync/oneshot/index.html), or combinators (like and_then etc). I can probably help you out with that if you have a specific idea in mind (:

What I currently propose to do is to remove the deadlock test and continue working on https://github.com/realcr/async_mutex/pull/8

**Comment by juchiast** [comment]
I see this deadlock while thinking about the new design for #8.

I'm really bothered about the fact that AcquireMutex takes the resource immediately in `acquire()`, this deadlock is an effect of it (I know that this deadlock is not very practical). Why do you think this is a correct behavior? Futures should do nothing unless polled, here it acctually do something.

Other thing:

With the fix in bf75932,  I noticed that we are using `Core::run()` the wrong way. According to the doc, `run()` wait for the passed future to complete, here we use it to wait for all the spawned futures too (`Core` also has `turn()`, which is said to be used with spawned futures).

Also, in `multiple` test (and other tests as well), we assume that spawned futures will be completed in order, which is true in this code, but that not how futures behave in general. Futures are free to be completed whenever they want, that is the point of asynchonous programming, right? This kind of task is usually done with a combinator, not a plain for loop.

My points are:
- We should reconsider the behavior of `acquire()`. It should do nothing unless polled.
- Rewrite tests in a more correct way.

These points are not related to deadlock anymore, we might open an issue to discuss.

**Comment by realcr** [comment]
> I'm really bothered about the fact that AcquireMutex takes the resource immediately in acquire()

@juchiast: I agree that this is somewhat strange behaviour.
Regarding the tests: If we assume that the method acquire() function adds a waiter immediately to the queue, then we can be sure about the order of the tasks. This is because the `async_mutex` itself enforces the order. This is true for all the current tests.

However, if we do somehow manage to change the implementation so that AcquireMutex does not take the resource immediately, then I agree that we should rewrite the tests, because we can not be sure about the order of the execution of the tasks any more.

Do you have any example where the behaviour of adding a waiter immediately when acquire() is called could be a problem? The only one I can think of is the case where someone runs acquire() but never runs the resulting Future. To avoid this, we could add a warning that requires using the resulting Future. (I think it is called must_use).

If you think that this behaviour is really a problem, I think that I might have some ideas about how to solve it. But first I think that we should find out if this is really a problem.

**Comment by juchiast** [comment]
>However, if we do somehow manage to change the implementation so that AcquireMutex does not take the resource immediately

I already done that in bf75932, I moved the code in `acquire()` into `poll()`.

**Comment by juchiast** [comment]
> Do you have any example where the behaviour of adding a waiter immediately when acquire() is called could be a problem?

I can think of one: you create some futures, store it somewhere or maybe move it to another function, then poll later. Now you have to poll in the exact order you created it. I'm not sure if this is a realistic one.

**Comment by realcr** [comment]
> I can think of one: you create some futures, store it somewhere or maybe move it to another function

Sounds reasonable to me. I didn't think about it. 

I'm always up for trying something interesting. If you can make this work, I probably vote for the new design.  What is the current state of your work on this? Are you stuck on something or are you still making progress?

I suspect we might need to take a handle to the Core at `AsyncMutex::new` if we want to make this new design work. I'm not fully sure about it though.

**Comment by juchiast** [comment]
> What is the current state of your work on this? Are you stuck on something or are you still making progress?

Did you miss my previous comment?

> I already done that in bf75932, I moved the code in acquire() into poll().

Anyway, I think that the code in bf75932 will work properly. You might want to have a look at it.

**Comment by realcr** [comment]
> Did you miss my previous comment?

No, sorry! I didn't realize I should look at https://github.com/realcr/async_mutex/commit/bf7593294f4bc1d92b80939e958d61f2e14210fc because the the tests didn't pass there.

It looks like it will work. Nicely done with the `NotPolled` state!
Only now I understand what you said about the tests. We should rewrite them if we want to work with the new design.

If you want consistent results, don't work with timers. Use `oneshot` instead. 
For example, assume that you have two futures: f1, f2. If you want to make sure f1 happens before f2, do the following (Pseudocode):

```
let (tx, rx)  = oneshot::new();
spawn(f1.and_then(tx.send());
spawn(rx.and_then(f2));
```

In other words, when f1 finishes a message is sent, and only when the message is received f2 will start running. Messages are the formal way of defining time in an async world (:

EDIT: I just realized `spawn(f1.and_then(f2))` will work just fine. We can keep the messaging for a more difficult problem (:

**Comment by juchiast** [comment]
@realcr Thanks for suggestion, I'm not sure how this is used to replace timeout.

Currently I'm using timeout to catch the deadlock. I run the Futures and set the timeout, if it fails to poll after 1 second, it will return `Err()`.

I don't think this can be done via message passing.

**Comment by realcr** [comment]
@juchiast : Try this line for `deadlock()`:

```rust
core.run(task0.and_then(task2).and_then(task1)).unwrap();
```
I didn't try to compile it, you might have to make some adjustments there. Please tell me if it works for you.

**Comment by juchiast** [comment]
@realcr No, it doesn't work. It just blocks the process forever.

The problem here is not executing futures in order.

Using:

``` rust
core.run(task0).unwrap();
core.run(task2).unwrap();
core.run(task1).unwrap();
```
is not different from:
``` rust
core.run(task0.and_then(|_| task2).and_then(|_| task1)).unwrap();
```

The problem is having a way to panic when there is deadlock. If there is deadlock, the above code will just hang forever. The purpose of `deadlock` test is to catch this deadlock, so it must have a way to panic when deadlock happens.

Using Timeout here is just like calling [Core::turn](https://docs.rs/tokio-core/0.1.17/tokio_core/reactor/struct.Core.html#method.turn) with appropriate `max_wait`.

**Comment by realcr** [comment]
> The problem is having a way to panic when there is deadlock. If there is deadlock, the above code will just hang forever

I understand now, you want that the test will fail, and not block forever. Don't worry about this. The test suite will report a failure if a test hangs for too long. It's not your test's job.

Write the code that fixes the test and commit it together with the test. As said earlier, I propose to avoid using timers in your tests, because they either make the tests suite slow, or create non deterministic tests. 

EDIT: I rather have a test that hangs forever if fails, than a test that sometimes passes and sometimes fails.
EDIT2: I'm on gitter if you want to discuss this with less latency (:

**Comment by juchiast** [comment]
@realcr okay. I replaced `core.spawn` with `core.run` in `simple` and `multiple` test.
In `deadlock` I use:

``` rust
core.run(task0).unwrap();
core.run(task2).unwrap();
core.run(task1).unwrap();
```

Should be okay to merge now.

**Review by realcr — APPROVED** [review]
Superb!
